metagpt/environment/werewolf_env/werewolf_ext_env.py

```diff
@@ -0,0 +1,274 @@
+        self.special_role_players = [
+            p for p in self.living_players if p not in self.werewolf_players + self.villager_players
+        ]
+
+        # init role state
+        self.roles_state = {player_name: RoleState.ALIVE for player_name in self.living_players}
+
+    @mark_as_readable
+    def init_game_setup(
+        self,
+        role_uniq_objs: list[object],
+        num_villager: int = 2,
+        num_werewolf: int = 2,
+        shuffle=True,
+        add_human=False,
+        use_reflection=True,
+        use_experience=False,
+        use_memory_selection=False,
+        new_experience_version="",
+        prepare_human_player=Callable,
+    ) -> tuple[str, list]:
+        role_objs = []
+        for role_obj in role_uniq_objs:
+            if str(role_obj) == "Villager":
+                role_objs.extend([role_obj] * num_villager)
+            elif str(role_obj) == "Werewolf":
+                role_objs.extend([role_obj] * num_werewolf)
+            else:
+                role_objs.append(role_obj)
+        if shuffle:
+            random.shuffle(len(role_objs))
+        if add_human:
+            assigned_role_idx = random.randint(0, len(role_objs) - 1)
+            assigned_role = role_objs[assigned_role_idx]
+            role_objs[assigned_role_idx] = prepare_human_player(assigned_role)  # TODO
+
+        players = [
+            role(
+                name=f"Player{i + 1}",
+                use_reflection=use_reflection,
+                use_experience=use_experience,
+                use_memory_selection=use_memory_selection,
+                new_experience_version=new_experience_version,
+            )
+            for i, role in enumerate(role_objs)
+        ]
+
+        if add_human:
+            logger.info(f"You are assigned {players[assigned_role_idx].name}({players[assigned_role_idx].profile})")
+
+        game_setup = ["Game setup:"] + [f"{player.name}: {player.profile}," for player in players]
+        game_setup = "\n".join(game_setup)
+
+        return game_setup, players
+
+    @mark_as_readable
+    def curr_step_instruction(self) -> dict:
+        step_idx = self.step_idx % len(STEP_INSTRUCTIONS)
+        instruction = STEP_INSTRUCTIONS[step_idx]
+        self.step_idx += 1
+        return instruction
+
+    @mark_as_writeable
+    def update_players_state(self, player_names: list[str], state: RoleState = RoleState.KILLED):
+        for player_name in player_names:
+            if player_name in self.roles_state:
+                self.roles_state[player_name] = state
+
+    @mark_as_readable
+    def get_players_status(self, player_names: list[str]) -> dict[str, RoleState]:
+        roles_state = {
+            player_name: self.roles_state[player_name]
+            for player_name in player_names
+            if player_name in self.roles_state
+        }
+        return roles_state
+
+    @mark_as_writeable
+    def wolf_kill_someone(self, player_name: str):
```

Comment:
These functions all seem to be permission-related, and at the same time, the characters that werewolves kill are actually inside ENV, and different characters have different sets of available APIs. If it seems difficult to deal with the issue of permissions with the current abstraction, what is a better way to solve it?

Reply:
add permission check when operation and update player state after voting or hunting.